### PR TITLE
redis: initial codec and proxy support

### PR DESCRIPTION
**Pull request by mattklein123**
This is an initial implementation of redis support in Envoy. It
includes a codec as well a simple proxy filter and "connection pool."
The connection pool is trivial and just makes a single connection. It
does no consistent hashing, op splitting, etc. However, this commit
is sufficient for Envoy to proxy to a single redis server and
encode/decode the entire conversation. All of the main interfaces
are in place.

To keep this change as self contained as possible there is quite a bit
missing that will be done in individual follow up changes:
  - stats
  - timeouts
  - consistent hashing
  - supported op checking and splitting where appropriate (e.g., MGET)
  - A real connection pool that supports hashing, and M connections to
    N upstreams
  - redis active HC and outlier detection

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @lyft/core-datastores

File: include/envoy/redis/codec.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  const std::string& asString() const;
+  int64_t& asInteger();
+  int64_t asInteger() const;
+
+  RespType type() const { return type_; }
+  void type(RespType type);
+
+private:
+  union {
+    std::vector<RespValue> array_;
```
Comment (mikramulhaq) [inline include/envoy/redis/codec.h]: Although union is more efficient, to keep it simple I'll avoid union here and just use three objects that will make destruction and construction super clean.

File: source/common/redis/codec_impl.cc [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    break;
+  }
+  }
+}
+
+void DecoderImpl::decode(Buffer::Instance& data) {
+  uint64_t num_slices = data.getRawSlices(nullptr, 0);
+  Buffer::RawSlice slices[num_slices];
+  data.getRawSlices(slices, num_slices);
+  for (Buffer::RawSlice& slice : slices) {
```
Comment (mikramulhaq) [inline source/common/redis/codec_impl.cc]: const Buffer::RawSlice& ?

File: source/common/redis/codec_impl.cc [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    parseSlice(slice);
+  }
+
+  data.drain(data.length());
+}
+
+void DecoderImpl::parseSlice(const Buffer::RawSlice& slice) {
+  char* buffer = reinterpret_cast<char*>(slice.mem_);
```
Comment (mikramulhaq) [inline source/common/redis/codec_impl.cc]: const char* ?

File: source/common/redis/codec_impl.cc [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+  uint64_t remaining = slice.len_;
+
+  while (remaining || state_ == State::ValueComplete) {
+    log_trace("parse slice: {} remaining", remaining);
```
Comment (mikramulhaq) [inline source/common/redis/codec_impl.cc]: log level? is it not going to be too much logging?

File: source/common/redis/codec_impl.cc [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+  }
+
+  data.drain(data.length());
+}
+
+void DecoderImpl::parseSlice(const Buffer::RawSlice& slice) {
+  char* buffer = reinterpret_cast<char*>(slice.mem_);
+  uint64_t remaining = slice.len_;
+
+  while (remaining || state_ == State::ValueComplete) {
```
Comment (mikramulhaq) [inline source/common/redis/codec_impl.cc]: 😎 
I understand this is hard part but can this code be bit simplified by breaking down to some functions please?

File: source/common/redis/conn_pool_impl.cc [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  encoder_->encode(request, encoder_buffer_);
+  connection_->write(encoder_buffer_);
+  return &pending_requests_.back();
+}
+
+void ClientImpl::onData(Buffer::Instance& data) {
+  try {
+    decoder_->decode(data);
+  } catch (ProtocolError&) {
+    connection_->close(Network::ConnectionCloseType::NoFlush);
```
Comment (mikramulhaq) [inline source/common/redis/conn_pool_impl.cc]: some log here?

File: include/envoy/redis/codec.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+
+namespace Redis {
```
Comment (mattklein123) [inline include/envoy/redis/codec.h]: This is not the standard we use.

File: include/envoy/redis/codec.h (hunk 1 above)
Comment (mattklein123) [inline include/envoy/redis/codec.h]: Performance is going to be an issue. The union is more complicated but it's worth it. We will keep it.

File: source/common/redis/codec_impl.cc (hunk 2 above)
Comment (mattklein123) [inline source/common/redis/codec_impl.cc]: fixed

File: source/common/redis/codec_impl.cc (hunk 3 above)
Comment (mattklein123) [inline source/common/redis/codec_impl.cc]: fixed

File: source/common/redis/codec_impl.cc (hunk 5 above)
Comment (mattklein123) [inline source/common/redis/codec_impl.cc]: Might refactor later. The code is easy to read w/ switch statement so will keep for now.

File: source/common/redis/codec_impl.cc (hunk 4 above)
Comment (mattklein123) [inline source/common/redis/codec_impl.cc]: log_trace is compiled out on release builds

File: source/common/redis/conn_pool_impl.cc (hunk 6 above)
Comment (mattklein123) [inline source/common/redis/conn_pool_impl.cc]: will add more logging, etc. later

File: include/envoy/redis/codec.h [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * All RESP types as defined here: https://redis.io/topics/protocol
+ */
+enum class RespType { Null, SimpleString, BulkString, Integer, Error, Array };
+
+/**
+ * A variant implementation of a RESP value optimized for peformance. A C++11 union is used for
+ * the underlying type so that no unnecessary allocations/constructions are needed.
+ */
+class RespValue {
```
Comment (ccaraman) [inline include/envoy/redis/codec.h]: General question about how we want to structure the code base: Is the intent for the headers under include/envoy/ to be interfaces?

Comment (mattklein123) [inline include/envoy/redis/codec.h]: In 99% of cases, yes. I'm will to compromise in certain cases mostly as it relates to performance. (We did something similar with HTTP headers). RespValue is guaranteed to become a performance bottleneck for the redis code. In order to use unions and allow vectors that contain actual RespValue vs. pointers, we need a concrete class and not an interface. So in this particular case I think it's worth it to have an exception to the general rule.

File: include/envoy/redis/codec.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+  std::vector<RespValue>& asArray();
+  const std::vector<RespValue>& asArray() const;
+  std::string& asString();
+  const std::string& asString() const;
+  int64_t& asInteger();
+  int64_t asInteger() const;
+
+  RespType type() const { return type_; }
+  void type(RespType type);
```
Comment (ccaraman) [inline include/envoy/redis/codec.h]: Please add comment for this method.

File: source/common/redis/conn_pool_impl.cc [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+ActiveRequest* InstanceImpl::makeRequest(const std::string&, const RespValue& value,
+                                         ActiveRequestCallbacks& callbacks) {
+  return tls_.getTyped<ThreadLocalPool>(tls_slot_).makeRequest(value, callbacks);
+}
+
+ActiveRequest* InstanceImpl::ThreadLocalPool::makeRequest(const RespValue& request,
+                                                          ActiveRequestCallbacks& callbacks) {
+  if (!client_) {
+    client_.reset(new ThreadLocalActiveClient(*this));
+    client_->client_ = parent_.client_factory_.create(parent_.cluster_name_, parent_.cm_);
```
Comment (ccaraman) [inline source/common/redis/conn_pool_impl.cc]: nit: rename one of the client_'s to be more descriptive?

File: source/common/redis/conn_pool_impl.cc [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+ClientPtr ClientImpl::create(const std::string& cluster_name, Upstream::ClusterManager& cm,
+                             EncoderPtr&& encoder, DecoderFactory& decoder_factory) {
+
+  Upstream::Host::CreateConnectionData data = cm.tcpConnForCluster(cluster_name);
+  if (!data.connection_) {
+    return nullptr;
+  }
+
+  std::unique_ptr<ClientImpl> client(new ClientImpl(std::move(encoder), decoder_factory));
```
Comment (mikramulhaq) [inline source/common/redis/conn_pool_impl.cc]: std::make_unique ?

File: test/common/redis/codec_impl_test.cc [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+  value.type(RespType::SimpleString);
+  value.asString() = "simple string";
+  encoder_.encode(value, buffer_);
+  EXPECT_EQ("+simple string\r\n", TestUtility::bufferToString(buffer_));
+  decoder_.decode(buffer_);
+  EXPECT_EQ(value, *decoded_values_[0]);
+  EXPECT_EQ(0UL, buffer_.length());
+}
+
+TEST_F(RedisEncoderDecoderImplTest, Integer) {
```
Comment (ccaraman) [inline test/common/redis/codec_impl_test.cc]: Please add test for max int we support.

Comment (ccaraman) [inline test/common/redis/codec_impl_test.cc]: And negative ints.

File: source/common/redis/codec_impl.cc [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+    encodeInteger(value.asInteger(), out);
+    break;
+  }
+}
+
+void EncoderImpl::encodeArray(const std::vector<RespValue>& array, Buffer::Instance& out) {
+  char buffer[32];
+  char* current = buffer;
+  *current++ = '*';
+  current += StringUtil::itoa(current, 31, array.size());
```
Comment (ccaraman) [inline source/common/redis/codec_impl.cc]: Is the math here wrong? Shouldn't this be 29? '*', '\r' and '\n'

File: source/common/redis/conn_pool_impl.cc (hunk 11 above)
Comment (mattklein123) [inline source/common/redis/conn_pool_impl.cc]: not in C++11

File: source/common/redis/codec_impl.cc (hunk 13 above)
Comment (mattklein123) [inline source/common/redis/codec_impl.cc]: At the time this has been called only 1 byte has been used, so it is correct.

File: test/common/redis/codec_impl_test.cc (hunk 12 above)
Comment (mattklein123) [inline test/common/redis/codec_impl_test.cc]: There are other tests that cover negative but I will add another test case.

Comment (mattklein123) [inline test/common/redis/codec_impl_test.cc]: I'm not really sure what other clients do about overflow in general. I will add a follow up on the GH issue to figure out how to deal with overflow in the codec.

Comment (mattklein123) [inline test/common/redis/codec_impl_test.cc]: I added tests for max/min int and added follow up to deal with overflow later.

File: source/common/redis/codec_impl.cc (hunk 13 above)
Comment (ccaraman) [inline source/common/redis/codec_impl.cc]: Right, max int takes up 21 chars.

**Pull request merged**